File: dashboard/src/api/exactProductOutput.ts
/**
 * Exact-product final-output client.
 *
 * For products with exact_product_composite_required the operator path is:
 *   1. GET policy / validate (fail closed before credits)
 *   2. Generate scene-only plate (NO product reference to Flow)
 *   3. POST compose-from-plate → final composite media only
 */
import { getAPI, postAPI } from "./client";

export interface ExactProductPolicy {
	product_id: string;
	product_display_name?: string | null;
	exact_product_composite_required: boolean;
	scene_only_required?: boolean;
	send_product_reference_to_flow?: boolean;
	canonical_valid?: boolean;
	canonical?: {
		schema_key?: string;
		source_sha256?: string;
		source_path?: string;
	} | null;
	error?: { code?: string; message?: string } | null;
	scene_only_prompt_block?: string;
	progress_stages?: string[];
	ok?: boolean;
}

export interface ExactComposeResult {
	ok: boolean;
	product_id: string;
	media_id: string;
	url: string;
	output_sha256: string;
	size_mb: number;
	status: string;
	truth_status: string;
	preview_sha_equals_saved_sha: boolean;
	lineage: {
		raw_plate_media_id?: string | null;
		raw_plate_sha256?: string;
		final_media_id?: string;
		final_output_sha256?: string;
		raw_plate_approvable?: boolean;
		truth_status?: string;
		transform?: Record<string, number>;
		qa?: Record<string, unknown>;
	};
	stages_completed?: string[];
}

export async function fetchExactProductPolicy(
	productId: string,
): Promise<ExactProductPolicy> {
	return getAPI<ExactProductPolicy>(`/api/exact-product/policy/${productId}`);
}

export async function validateExactProduct(
	productId: string,
): Promise<ExactProductPolicy> {
	return postAPI<ExactProductPolicy>(
		`/api/exact-product/validate/${productId}`,
		{},
	);
}

export async function buildExactSceneOnlyPrompt(
	productId: string,
	prompt: string,
): Promise<{
	product_id: string;
	exact_product_composite_required: boolean;
	prompt: string;
	send_product_reference_to_flow: boolean;
}> {
	return postAPI(`/api/exact-product/scene-only-prompt`, {
		product_id: productId,
		prompt,
	});
}

export async function composeExactFromPlate(input: {
	product_id: string;
	background_media_id?: string;
	background_local_path?: string;
	lane?: "studio" | "poster" | "product_only_hero";
	job_id?: string;
}): Promise<ExactComposeResult> {
	return postAPI<ExactComposeResult>("/api/exact-product/compose-from-plate", {
		product_id: input.product_id,
		background_media_id: input.background_media_id ?? null,
		background_local_path: input.background_local_path ?? null,
		lane: input.lane ?? "studio",
		job_id: input.job_id ?? null,
	});
}

File: agent/api/exact_product_output.py
"""Exact product final-output API.

Fail-closed compose path for products with exact_product_composite_required.
Raw Google Flow plates stay internal; only the deterministic composite is
returned as a final media artifact.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, ConfigDict, Field

from agent.services.exact_product_compositor_service import ExactProductCompositeError
from agent.services import exact_product_final_output_service as svc

router = APIRouter(prefix="/exact-product", tags=["exact-product-output"])


class ComposeFromPlateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str = Field(..., min_length=1)
    background_media_id: str | None = None
    background_local_path: str | None = None
    lane: str = "studio"
    job_id: str | None = None


class SceneOnlyPromptRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str = Field(..., min_length=1)
    prompt: str = Field(..., min_length=1)


def _http(exc: ExactProductCompositeError) -> HTTPException:
    return HTTPException(
        status_code=exc.status_code,
        detail={"error": exc.code, "detail": exc.message},
    )


@router.get("/policy/{product_id}")
async def get_exact_product_policy(product_id: str):
    """Pre-generation policy + canonical validity (no credits)."""
    try:
        return await svc.get_policy_for_product(product_id)
    except ExactProductCompositeError as exc:
        raise _http(exc) from exc


@router.post("/validate/{product_id}")
async def validate_exact_product(product_id: str):
    """Fail-closed pre-credit canonical check."""
    try:
        policy = await svc.get_policy_for_product(product_id)
        if not policy.get("exact_product_composite_required"):
            return {**policy, "ok": True}
        if not policy.get("canonical_valid"):
            err = policy.get("error") or {}
            raise ExactProductCompositeError(
                err.get("code") or "CANONICAL_PRODUCT_SOURCE_INVALID",
                err.get("message") or "canonical invalid",
                status_code=422,
            )
        return {**policy, "ok": True}
    except ExactProductCompositeError as exc:
        raise _http(exc) from exc


@router.post("/scene-only-prompt")
async def build_scene_only_prompt(req: SceneOnlyPromptRequest):
    """Augment operator/poster prompt for exact-policy scene plates."""
    try:
        policy = await svc.get_policy_for_product(req.product_id)
        if not policy.get("exact_product_composite_required"):
            return {
                "product_id": req.product_id,
                "exact_product_composite_required": False,
                "prompt": req.prompt,
                "send_product_reference_to_flow": True,
            }
        if policy.get("exact_product_composite_required") and not policy.get(
            "canonical_valid"
        ):
            err = policy.get("error") or {}
            raise ExactProductCompositeError(
                err.get("code") or "CANONICAL_PRODUCT_SOURCE_INVALID",
                err.get("message") or "canonical invalid",
                status_code=422,
            )
        return {
            "product_id": req.product_id,
            "exact_product_composite_required": True,
            "prompt": svc.build_scene_only_prompt(req.prompt),
            "send_product_reference_to_flow": False,
            "scene_only_prompt_block": policy.get("scene_only_prompt_block"),
        }
    except ExactProductCompositeError as exc:
        raise _http(exc) from exc


@router.post("/compose-from-plate")
async def compose_from_plate(req: ComposeFromPlateRequest):
    """Insert canonical cutout onto a retrieved scene plate; return final only."""
    try:
        return await svc.compose_final_for_product(
            product_id=req.product_id,
            background_media_id=req.background_media_id or "",
            background_local_path=req.background_local_path or "",
            lane=req.lane or "studio",
            job_id=req.job_id,
        )
    except ExactProductCompositeError as exc:
        raise _http(exc) from exc
